docs/proposals/2022-06-27_Auth_Proposal.md

```diff
@@ -0,0 +1,62 @@
+# Auth Proposal
+
+# Problem
+
+The [Signing Service](https://github.com/WPMedia/signing-service) was created to provide Photo Center & PageBuilder HMAC tokens which allow the use of images on the frontend. Without a field to store these tokens in, users would have to call the Signing Service for each image they want to display.
+
+# Proposal
+
+Add a field `auth` which contains a mapping of integer -> HMAC token, where the integer is the version of the secret key in Signing Service. The secret key is stored in SSM and is only known by Signing Service and in Akamai/Edge. 
+
+When an image is requested on a customer site, a signed token must be passed in a query parameter `auth`. The customer's Akamai property will sign its own token and compare it to the one passed in `auth`.
+
+For displaying Photo Center images, Photo Center will provide the `auth` object in all image ANS, including signed tokens with all available secret versions from Signing Service.
+
+For displaying external images, PageBuilder (or another user) must call the Signing Service with the full URL of the image to be used as the token in `auth`.
+
+## Key
+
+There may be more than 1 currently-enabled secret version at a given time for an organization. Users will know which version they're using, so the `auth` object's keys should be the secret version integer from Signing Service's SSM storage.
+
+## Value
+
+The values corresponding to each secret version integer will be the HMAC token produced by Signing Service for this object.
+
+# Example
+
+```
+{
+    "_id": "ABCDEFGHIJKLMNOPQRSTUVWXYZ",
+    "additional_properties": {
+        ...
+    },
+    "auth": {
+        "1": "c86ec5d13cb475fb790d627c1f9754c4ebb90718d4daaa1d027731e2075b9b05",
+        "2": "8765d450481157336afff9e8e57f67ce14a7d627e50e7c652f3a6360243e2c6b"
+    },
+    "caption": "This is a caption.",
+    ...
+    "subtitle": "This is a subtitle.",
+    "type": "image",
+    "version": "0.10.9"
+}
+```
+
+# Concerns
+
+> Aren't you exposing a token in ANS?
+
+It is just the signed token, so there is no risk of exposing the secret key, and the token is only valid for that single image.
+
+> What prevents someone from reading the token from ANS embedded on a website and using it?
+
+We decided to accept this risk now that we aren't worried about the performance of Thumbor in ECS. Akamai is much more durable and has protections in place for DDOS & other exploits.
```

Comment:
I see.  This is security through obfuscation which is marginally better than no security at all.

For token based security like this, being able to revoke a token is typically a requirement as eventually a bad actor will figure out our obfuscation pattern.  How does token revocation and replacement work for this system?  Simply create a new versioned token and deactivate the compromised one?

My concern here is that once a bad actor understands our obfuscation pattern then this auth system is effectively broken.  Most client-side auth systems suffer from this drawback and this drawback is hedged against by using secure granular tokens with limited access rights and a token revocation/replacement process.  

Can you document the token revocation/replacement process in this proposal?
Can you document the limitations of this auth system?

Lastly, we should get AppSec on this PR once the proposal is finalized.

Reply:
@ibaker thanks for the in-depth explanation of the auth system this ANS change is interacting with as the context helps explain most of the why of this PR.